histomicstk/utils/sample_pixels.py

```diff
@@ -0,0 +1,117 @@
+import large_image
+import numpy as np
+import scipy
+
+from .simple_mask import simple_mask
+
+
+def sample_pixels(slide_path, magnification, percent, tile_size,
+                  mapping_mag=1.25, min_coverage=0.1):
+    """Generates a sampling of pixels from a whole-slide image.
+
+    Useful for generating statistics or Reinhard color-normalization or
+    adaptive deconvolution. Uses mixture modeling approach to focus
+    sampling in tissue regions.
+
+    Parameters
+    ----------
+    slide_path : str
+        path and filename of slide.
+    magnification : double
+        Desired magnification for sampling (defaults to native scan
+        magnification).
+    percent : double
+        Percentage of pixels to sample. Must be in the range [0, 1].
+    tile_size : int
+        tile_size size used in sampling high-resolution image.
+    mapping_mag: double, optional
+        low resolution magnification. Default value = 1.25.
+    min_coverage: double, optional
+        minimum percent of tile covered by tissue to be included in sampling.
+        Ranges between [0,1). Default value = 0.1.
+
+    Returns
+    -------
+    Pixels : array_like
+        A 3xN matrix of RGB pixel values sampled from the whole-slide.
+
+    See Also
+    --------
+    histomicstk.preprocessing.color_normalization.reinhard,
+    histomicstk.preprocessing.color_deconvolution.SparseColorDeconvolution
+    """
+
+    ts = large_image.getTileSource(slide_path)
+
+    # get enitre whole-silde image at low resolution
+    scale_lowres = {'magnification': mapping_mag}
+    im_lowres, _ = ts.getRegion(
+        format=large_image.tilesource.TILE_FORMAT_NUMPY,
+        scale=scale_lowres
+    )
+
+    # compute foreground mask of whole-slide image at low-res
+    fgnd_mask_lowres = simple_mask(im_lowres)
+
+    # generate sample pixels
+    sample_pixels = []
+
+    scale_highres = {'magnfication': magnification}
+
+    i = 0
+
+    for tile in ts.tileIterator(
+            scale=scale_highres,
+            format=large_image.tilesource.TILE_FORMAT_NUMPY):
+
+        # get current tile image
+        im_tile = tile['tile'][:, :, :3]
+
+        # get current region in base_pixels
+        rgn_hres = {'left': tile['gx'], 'top': tile['gy'],
+                    'width': tile['gwidth'], 'height': tile['gheight'],
+                    'units': 'base_pixels'}
```

Comment:
Just an idea: Maybe it is better to encapsulate the region properties `left top width height` into a dict like `tile['mag_region']` and tile['global_region']?